Password strength indicator

narickmann opened the pull request:
This fixes #3215.
Simple strength bar, checks if the password contains at least one uppercase letter, one lowercase letter, one number, one special character and is at least 8 characters long. (Colors of strength-bar: red, orange, yellow, green, darkgreen)

![strength-bar-1](https://user-images.githubusercontent.com/93541078/148920412-afaf0aee-d642-4811-819b-5ee320a6a217.png)

![strength-bar-4](https://user-images.githubusercontent.com/93541078/148920514-6ce9da01-b48c-4865-a061-f7c96eb3db93.png)

![strength-bar-6](https://user-images.githubusercontent.com/93541078/148920430-00472132-8f47-4f5d-8183-f90cd410565a.png)


### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [x] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

Commit:
Password strength indicator

This fixes #3215.
Simple strength bar, checks if the password contains at least one uppercase letter, one lowercase letter, one number, one special character and is at least 8 characters long.

gregorydlogan commented:
Quick comments:
- The colour contrast for the yellow line is hard to make out
- The green colour in the last image should probably match the rest of the UI branding.

lkiesow commented:
An additional textual indicator (weak, good, strong) would also be helpful

Commit:
Password strength indicator

Changed the colors (now red, darkorange, gold, green and blue #388ed5) and added a textual indicator.
Tests should now run without failures.

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,93 @@
+                pw_strength = 'Very Strong';
+              }
+              document.getElementById('bar').style = bar;
+              document.getElementById('pw').innerHTML = pw_strength;
```
Safer and should work the same:

```suggestion
              document.getElementById('pw').innerText = pw_strength;
```

narickmann commented:
Yellow is now darker (gold) and darkgreen is now blue. Also added a textual indicator. Pls let me know, if the colors are okay or if you want me to change them or the textual indicator.

![strength-bar-1](https://user-images.githubusercontent.com/93541078/149146722-c5da1de5-bc4d-4c59-8b1f-72bc0da9c3b9.png)
![strength-bar-2](https://user-images.githubusercontent.com/93541078/149146728-f1f7bea9-ef33-4e59-b83a-c5eb89e56452.png)
![strength-bar-3](https://user-images.githubusercontent.com/93541078/149146734-4dffab3b-ee2b-4eae-b297-1869279eafbe.png)
![strength-bar-4](https://user-images.githubusercontent.com/93541078/149146741-4a060cd6-d8cd-45d6-b40d-81dca17b450b.png)
![strength-bar-5](https://user-images.githubusercontent.com/93541078/149146745-5010ace8-c73e-42c6-bd40-fc881e0c9ddb.png)
![strength-bar-6](https://user-images.githubusercontent.com/93541078/149146748-8831b899-098f-4b73-9a76-2b5be57b1666.png)

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,93 @@
+                isValid(pw && /[a-z]/.test(pw)) +
+                isValid(pw && /\d/.test(pw)) +
+                isValid(pw && /(?=.*\W)/.test(pw)) +
+                isValid(pw && /^.{8,}$/.test(pw));
```
This just awards a single point to all passwords longer than 8 characters (identical to `pw.length >= 8`).
How about awarding a point for each 8 unique characters instead? That way, a very long password with lots of different characters gets extra points, while just repeating a single character 8 times gets no points.
Something like:

```suggestion
                evalLength(pw);
            function evalLength(password) {
                let unique = '';
                for (const c of password) {
                    if (u.indexOf(c) < 0) {
                        unique += c;
                    }
                }
                return Math.floor(unique.length / 8);
            }
```

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,93 @@
+            }
+
+            function setProgBar(validRules) {
+              if (validRules == 1) {
```
Maybe additionally check against the most commonly used bad passwords:

```suggestion
// bad passwords from https://en.wikipedia.org/wiki/List_of_the_most_common_passwords
// plus Opencast's default password
const bad_passwords = ["0", "111111", "1111111", "123", "123123", "123321",
	"1234", "12345", "123456", "1234567", "12345678", "123456789", "1234567890",
	"12345679", "123qwe", "18atcskd2w", "1q2w3e", "1q2w3e4r", "1q2w3e4r5t",
	"3rjs1la7qe", "555555", "654321", "666666", "7777777", "888888",
	"987654321", "aa12345678", "abc123", "admin", "dragon", "Dragon", "google",
	"iloveyou", "Iloveyou", "lovely", "Monkey", "mynoob", "password",
	"password1", "password123", "princess", "qwerty", "qwerty123", "qwertyuiop",
	"Qwertyuiop", "welcome", "zxcvbnm", "opencast" ]

              if (bad_passwords.indexOf(pw) >= 0) {
                  // This is one of the most commonly used (bad) passwords
                  bar = 'background:black; width:10%;';
                  pw_strength = 'Bad password';
              }
              else if (validRules == 1) {
```

lkiesow reviewed (changes requested):
I don't think I would necessarily agree with the algorithm:

```
yzvhcklsvojttgtpbljurxmgcrdkdvvopncywiunsdszxsafltczxbrohmafdgnvmzghhmjcalynbmxlemitrrnyqnkaoliesadchhrpckwslaxcksvlgjabaetleqxo
-> Weak
```

while

```
aA1
-> Strong
```

Now, we don't need to (and never will be) perfect, but maybe just factor in the length.
I'll add a few suggestions for that in the code below.

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,93 @@
+                bar = 'background:green; width:80%;';
+                pw_strength = 'Strong';
+              }
+              else if (validRules == 5) {
```
Using my suggestion, this check also would need a slight adjustment:

```suggestion
              else if (validRules >= 5) {
```

Commit:
Password strength indicator

Changed algorithm. Added list of bad passwords.

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,140 @@
+              // for every unused rule/requirement = -5 points
+              var usedRules = (rule - 5) * 5;
+
+              var uniqueChars = (pw.split('').sort().join('').replace(/(.)\1+/g, '').length) * 2;
```
This will actually count unique characters only. Meaning that e.g. `aaa` would count as `0` and not as `1`. This is probably not what you intended. What you want is probably:

```suggestion
              var uniqueChars = new Set(pw).size * 2;
```

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,140 @@
+                bar_width = 'width:' + 0 + '%;';
+                bar_text = 'Bad password';
+              }
+
```
Instead of repeating this in every condition above, you could set this once here:

```suggestion
              bar_width = 'width:' + strength + '%;';
```

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,140 @@
+                requirements(pw && /[A-Z]/.test(pw)) +
+                requirements(pw && /[a-z]/.test(pw)) +
+                requirements(pw && /\d/.test(pw)) +
+                requirements(pw && /(?=.*\W)/.test(pw)) +
```
I'm actually not entirely sure but isn't this just identical to

```suggestion
                requirements(pw && /\W/.test(pw)) +
```

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,140 @@
+              var symbol = (pw.length - pw.replace(/\W/g, '').length) * 6;
+
+              var strength = usedRules + uniqueChars + pw_length + lowerCase + upperCase + number + symbol;
+              return strength;
```
Let's assume we chose the password `a`.
The password `a`…
- is not in the list of bad passwords. Thus, we continue to the ranking.
- does not match 4 out of the 5 rules, giving us a starting score of `-20`.
- gets `2` points for having one unique character.
- gets `4` points for length.
- gets `2` points for lowercase character length.

We end up with `-20 + 2 + 4 + 2 = -12`.
But `-12` does not match any of your conditions below.
You probably want all of these to be very weak:

```suggestion
              return Math.max(1, usedRules + uniqueChars + pw_length + lowerCase + upperCase + number + symbol);
```

lkiesow commented on modules/admin-ui-frontend/app/scripts/shared/directives/passwordStrengthDirective.js:
```diff
@@ -0,0 +1,140 @@
+              if (strength >= 90) {
+                bar_color = 'background:#388ed6;';
+                bar_width = 'width:' + strength + '%;';
+                bar_text = 'Very strong';
```
These strings should probably be translated:

https://github.com/opencast/opencast/blob/cdfd04c33725fc71c3d682859fc863fe6937b668/modules/admin-ui-frontend/app/scripts/shared/services/restServiceMonitor.js#L72

Commit:
Password-strength indicator

This fixes #3215
Implemented proposed changes and translation for textual indicator.

lkiesow reviewed (changes requested):
Generally looks good.

The only thing missing is the source translation. [lang-en_US.json](https://github.com/opencast/opencast/blob/52b63cef0beb180a64c51e67f5db68ddc21517b9/modules/admin-ui-frontend/resources/public/org/opencastproject/adminui/languages/lang-en_US.json) is used as source of the translation. All other translation files are automatically overwritten by whatever was translated on Crowding. You modified `en_GB` and `de_DE`. Both will be overwritten. You can probably just copy over the changes from `en_GB` to `en_US`.

gregorydlogan commented:
@naweidner are you moving forward with this?

Commit:
Password strength indicator

Added source translation in adminui/languages/lang-en_US.json

narickmann commented:
Hopefully I've adressed through all the remaining open points.
